File: backend/src/services/ai/interviewPredictor.js
/**
 * Interview Question Predictor
 *
 * Generates likely behavioral and technical questions
 * grounded in the resume's actual content.
 */

import { llmCompleteJSON } from './llmClient.js';

/**
 * Predict interview questions based on resume content.
 * @param {string} resumeText
 * @param {object} parsedJson
 * @returns {{ questions: Array<{ type: string, question: string, context: string, tip: string }> }}
 */
export async function predictInterviewQuestions(resumeText, parsedJson) {
  const result = await llmCompleteJSON(
    `You are an experienced technical interviewer and hiring manager. Based on the candidate's resume, predict the most likely interview questions they would face.

Rules:
- Generate questions that are DIRECTLY grounded in the resume's content
- Include a mix of behavioral (STAR-format) and technical questions
- Reference specific projects, technologies, or experiences from the resume
- For each question, explain WHY an interviewer would ask it (what they're probing)
- Include a brief tip on how to answer well
- Generate 8-12 questions total`,

    `RESUME:
${resumeText?.slice(0, 3000)}

Respond with a JSON object: {
  "questions": [
    {
      "type": "behavioral" or "technical" or "situational",
      "question": "the interview question",
      "context": "which part of the resume prompted this question",
      "tip": "brief advice on how to answer well"
    }
  ]
}`,
    { maxTokens: 2500 }
  );

  if (!result || !result.questions) {
    // Graceful fallback: generate generic but useful questions from resume sections
    return {
      questions: generateFallbackQuestions(parsedJson),
    };
  }

  return result;
}

/**
 * Generate basic interview questions without AI, based on resume structure.
 */
function generateFallbackQuestions(parsedJson) {
  const questions = [];
  const sections = parsedJson?.sections || {};

  if (sections.experience) {
    questions.push({
      type: 'behavioral',
      question: 'Tell me about a challenging project you worked on and how you overcame obstacles.',
      context: 'Work experience section',
      tip: 'Use the STAR format: Situation, Task, Action, Result. Be specific about your contribution.',
    });
    questions.push({
      type: 'behavioral',
      question: 'Describe a time you had to work with a difficult team member.',
      context: 'Team collaboration experience',
      tip: 'Focus on the positive outcome and what you learned.',
    });
  }

  if (sections.skills) {
    questions.push({
      type: 'technical',
      question: 'Walk me through your experience with the technologies listed in your skills section.',
      context: 'Technical skills',
      tip: 'Rate your proficiency honestly and provide specific project examples for your strongest skills.',
    });
  }

  if (sections.education) {
    questions.push({
      type: 'behavioral',
      question: 'How has your education prepared you for this role?',
      context: 'Education section',
      tip: 'Connect specific coursework or projects to the job requirements.',
    });
  }

  questions.push({
    type: 'situational',
    question: 'Where do you see yourself in 5 years?',
    context: 'Career trajectory',
    tip: 'Show ambition while aligning with the company\'s growth opportunities.',
  });

  questions.push({
    type: 'behavioral',
    question: 'What\'s your greatest professional achievement and why?',
    context: 'Overall resume',
    tip: 'Pick an achievement with measurable impact and explain the process, not just the result.',
  });

  return questions;
}

File: backend/app/services/ml/interview_predictor.py
"""
ResumeIQ — Interview Question Predictor (Local NLP Engine)
Generates tailored behavioral (STAR), technical probing, and situational interview questions
grounded directly in the candidate's parsed resume content and skills.
100% offline, zero external API keys.
"""

from typing import Dict, Any, List

def predict_interview_questions(resume_text: str, parsed_json: Dict[str, Any] = None) -> Dict[str, List[Dict[str, str]]]:
    """Generate grounded interview questions based on resume content, skills, and experience."""
    sections = (parsed_json or {}).get("sections", {}) or {}
    questions: List[Dict[str, str]] = []

    # 1. Experience-grounded questions
    exp_section = sections.get("experience", {})
    bullets = exp_section.get("bullets", []) if isinstance(exp_section, dict) else []
    
    if bullets:
        # Use first bullet for project challenge
        b1 = bullets[0]
        questions.append({
            "type": "behavioral",
            "question": f"In your experience, you mentioned '{b1[:100]}...'. Can you walk me through the key technical obstacles you faced and how you overcame them?",
            "context": "Work experience achievement",
            "tip": "Use the STAR method (Situation, Task, Action, Result). Emphasize your personal contribution and trade-off decisions.",
        })
        if len(bullets) > 1:
            b2 = bullets[1]
            questions.append({
                "type": "technical",
                "question": f"Regarding your work on '{b2[:100]}...', what architecture or methodology did you select and what were the alternatives considered?",
                "context": "System design and implementation",
                "tip": "Explain system trade-offs, scalability considerations, and metrics used to evaluate success.",
            })
    else:
        questions.append({
            "type": "behavioral",
            "question": "Tell me about a complex project you worked on recently. What was your role and what measurable outcome was achieved?",
            "context": "Professional experience",
            "tip": "Frame your answer with the STAR format: highlight the Situation, Task, Action, and quantified Result.",
        })

    # 2. Skills-grounded questions
    skills_sec = sections.get("skills", {})
    skills_content = skills_sec.get("content", "") if isinstance(skills_sec, dict) else ""
    if not skills_content and resume_text:
        # extract words from skills if present
        skills_content = resume_text

    lower_text = (skills_content + " " + resume_text).lower()

    if any(k in lower_text for k in ["react", "next", "vue", "frontend", "typescript", "javascript"]):
        questions.append({
            "type": "technical",
            "question": "How do you optimize state management, component re-rendering, and page load performance in modern web applications?",
            "context": "Frontend & JavaScript stack",
            "tip": "Discuss memoization, code splitting, virtual DOM/server components, and Core Web Vitals profiling.",
        })

    if any(k in lower_text for k in ["python", "fastapi", "django", "node", "backend", "api", "rest"]):
        questions.append({
            "type": "technical",
            "question": "How do you design scalable RESTful/gRPC APIs with proper authentication, rate limiting, and database query optimization?",
            "context": "Backend & API engineering",
            "tip": "Cover indexing, connection pooling, async concurrency, idempotency, and defense-in-depth security.",
        })

    if any(k in lower_text for k in ["sql", "postgres", "mongodb", "database", "redis"]):
        questions.append({
            "type": "technical",
            "question": "Describe a scenario where a database query or schema bottlenecked system performance. How did you diagnose and resolve it?",
            "context": "Database & Data Layer",
            "tip": "Discuss EXPLAIN ANALYZE, indexing strategies, caching layers, and schema normalization trade-offs.",
        })

    if any(k in lower_text for k in ["docker", "k8s", "kubernetes", "aws", "gcp", "ci/cd", "devops"]):
        questions.append({
            "type": "technical",
            "question": "Walk me through your end-to-end CI/CD and container orchestration strategy for zero-downtime production deployments.",
            "context": "DevOps & Cloud Infrastructure",
            "tip": "Highlight automated canary/blue-green deployments, health probes, rollback mechanisms, and secret management.",
        })

    # 3. Collaboration, Conflict & Leadership
    questions.append({
        "type": "behavioral",
        "question": "Describe a situation where you had a strong technical disagreement with a team member or stakeholder. How did you navigate the discussion to reach consensus?",
        "context": "Cross-functional collaboration",
        "tip": "Focus on data-driven reasoning, active listening, psychological safety, and aligning on product goals.",
    })

    # 4. Situational & Career
    questions.append({
        "type": "situational",
        "question": "If you inherited a legacy codebase with minimal documentation and tight deadlines, how would you prioritize refactoring versus delivering new features?",
        "context": "Engineering judgment & prioritization",
        "tip": "Demonstrate pragmatic balance: introduce automated safety tests first, decouple critical paths iteratively, and communicate risks.",
    })

    questions.append({
        "type": "situational",
        "question": "Where do you see yourself technically and professionally in the next 2-3 years, and how does this role fit your career trajectory?",
        "context": "Long-term vision & motivation",
        "tip": "Align your personal growth goals with the team's engineering challenges and business impact.",
    })

    return {"questions": questions}
